main/src/ca/uwaterloo/flix/language/phase/Parser.scala

```diff
@@ -502,8 +502,12 @@ class Parser(val source: Source) extends org.parboiled2.Parser {
       SP ~ keyword("discard") ~ WS ~ Expression ~ SP ~> ParsedAst.Expression.Discard
     }
 
+    def ForEachGenerator: Rule1[ParsedAst.ForEachGenerator] = rule {
+      SP ~ Pattern ~ WS ~ keyword("<-") ~ WS ~ Expression ~ SP ~> ParsedAst.ForEachGenerator
+    }
+
     def ForEach: Rule1[ParsedAst.Expression.ForEach] = rule {
-      SP ~ keyword("foreach") ~ optWS ~ "(" ~ optWS ~ Pattern ~ WS ~ keyword("<-") ~ WS ~ Expression ~ optWS ~ ")" ~ optWS ~ Expression ~ SP ~> ParsedAst.Expression.ForEach
+      SP ~ keyword("foreach") ~ optWS ~ "(" ~ optWS ~ oneOrMore(ForEachGenerator).separatedBy(optWS ~ ";" ~ optWS) ~ optWS ~ ")" ~ optWS ~ Expression ~ SP ~> ParsedAst.Expression.ForEach
```

Comment:
Do we want to the last whitespace to be hard or is it ok to write `foreach(...)123+2`?

Reply:
I personally think we shouldn't be too restrictive on the white space unless there's a really good reason to.

Some C programmers for example write their stuff without white space hehe :wink: